### PR TITLE
feat: getCanonicalTopicOverlap server action

### LINKED ISSUE #395: feat: getCanonicalTopicOverlap server action
Part of #378

## Context

Adds a server action `getCanonicalTopicOverlap` that returns a structured indicator describing how many of a user's listened/saved episodes share canonical topics with a given episode. This is the canonical-topic-level analog to ADR-034's category-level `getTopicOverlapIndicator`. Unlike categories, canonical topics are entity-resolved — "Opus 4.7", "Claude Opus 4.7", and "Anthropic's new Opus" all match each other — so the resulting signal is far more precise.

The action is consumed by C2 (UI integration on episode cards, summary display, episode detail page). It must preserve the category-level fallback so episodes without canonical-topic links continue to show the existing ADR-034 indicator.

**Spec:** `.dev/pm/specs/2026-04-25-canonical-topics-foundation.md` (Approved)

## Complexity Hint

**Subsystems touched:** New server action, new query helper in `src/lib/topic-overlap.ts`
**Independent stories:** 1 (single server action with one helper backing it)
**Structural changes needed:** No (additive)
**Scope note:** Pure server-side data layer; UI integration ships in C2.

## Task

1. Add `computeCanonicalTopicOverlap(userId, episodeId): Promise<CanonicalOverlapResult>` to `src/lib/topic-overlap.ts` alongside the existing `computeTopicOverlap` (category version):
   - Reads canonical topics linked to the target episode via `episode_canonical_topics` + `canonical_topics WHERE status='active'`.
   - For each canonical topic, count how many of the user's listened OR saved episodes are also linked to that canonical (excluding the target episode itself).
   - Pick the most prominent indicator: the canonical with the highest count; fall back to "new" if all have count 0.
   - Result shape: `{ kind: 'new'; topicLabel: string; topicId: number } | { kind: 'repeat'; count: number; topicLabel: string; topicId: number } | null` (`null` when no canonical topics exist on the target).
2. Add server action `getCanonicalTopicOverlap(episodeId): Promise<ActionResult<CanonicalOverlapResult>>` to `src/app/actions/topics.ts`:
   - Auth-gated via Clerk; reads `userId` from session.
   - Calls `computeCanonicalTopicOverlap(userId, episodeId)`.
   - Returns `{ ok: true, data: result }` or `{ ok: false, error: ... }`.
3. Add a batch variant `getCanonicalTopicOverlaps(episodeIds: number[]): Promise<ActionResult<Record<number, CanonicalOverlapResult>>>` for efficient list-page fetching.
4. The `kind` selection logic:
   - `'new'` → no other episode in the user's history shares any canonical topic (count=0 for all topics, or topics exist on target but user has no overlap).
   - `'repeat'` → at least one canonical has count ≥ 1 in the user's history; pick the max-count canonical for the indicator label.
   - `null` → target episode has no canonical topics linked (let the caller fall back to category indicator).
5. Tests:
   - Unit tests for `computeCanonicalTopicOverlap` with seeded data: target with no canonicals (returns null), target with 2 canonicals where user has 0 in history (returns 'new'), target with canonical the user has 3 episodes for (returns 'repeat' with count=3 and the right label).
   - Unit tests for the server action including auth rejection.
   - Unit tests for the batch variant.

## Acceptance Criteria

- [ ] VERIFY: `src/lib/topic-overlap.ts` exports `computeCanonicalTopicOverlap(userId, episodeId)` returning `{ kind: 'new' | 'repeat'; ... } | null`.
- [ ] VERIFY: Returns `null` when the target episode has no canonical topics linked.
- [ ] VERIFY: Returns `{ kind: 'new', topicLabel, topicId }` when target has canonical topics but user has 0 overlap.
- [ ] VERIFY: Returns `{ kind: 'repeat', count, topicLabel, topicId }` when at least one canonical overlaps; selects the canonical with the highest user-episode count.
- [ ] VERIFY: Excludes the target episode itself from overlap counts.
- [ ] VERIFY: Considers both listened (`isListened=true`) AND saved episodes for the user's history.
- [ ] VERIFY: Filters out merged canonicals (`status='active'` only).
- [ ] VERIFY: `src/app/actions/topics.ts` exports `getCanonicalTopicOverlap(episodeId)` (auth-gated) returning `ActionResult<CanonicalOverlapResult>`.
- [ ] VERIFY: `getCanonicalTopicOverlaps(episodeIds)` batch variant returns a map keyed by episodeId.
- [ ] VERIFY: Unauthenticated callers receive `ActionResult.error` with status `unauthorized`.
- [ ] VERIFY: Tests in `src/lib/__tests__/topic-overlap.canonical.test.ts` and `src/app/actions/__tests__/topics.test.ts` cover all paths.
- [ ] VERIFY: `bun run lint`, `bun run test`, `bun run build` pass.

### Files to Create

- `src/lib/__tests__/topic-overlap.canonical.test.ts`

### Files to Modify

- `src/lib/topic-overlap.ts` — add `computeCanonicalTopicOverlap`.
- `src/app/actions/topics.ts` — add `getCanonicalTopicOverlap` and the batch variant.
- `src/app/actions/__tests__/topics.test.ts` — extend tests.

### Approach

1. **Core query.** A single SQL aggregation: for the target episode, get its canonical-topic ids; for each, count episodes in the user's history (listened or saved) that share that canonical. Drizzle relational + raw `sql` aggregation works; alternatively two queries (target's canonicals → counts per canonical) is fine.
2. **Indicator selection.** Take the max-count canonical; if all counts are 0, return 'new' with the most-prominent canonical (highest `coverage_score` from junction).
3. **Server action.** Auth via Clerk; call the helper; return structured.
4. **Batch variant.** Avoid N+1 — single SQL pulling overlaps for all episodes in one shot, then bucketize by episodeId.
5. **Tests.** Seed users + episodes + listened/saved + canonical_topics + junction; assert outputs.

### Patterns to Follow

- Existing `src/lib/topic-overlap.ts` (category version) — mirror its function signature and result shape conventions.
- ADR-034 for the category overlap logic that this complements.
- Existing server-action patterns: `"use server"` + `auth()` + structured `ActionResult`.

### Constraints

- ONE indicator per episode — the action returns a single structured result. The caller picks `repeat` when count ≥ 1, `new` when count = 0, `null` when no canonicals exist.
- `status='active'` filter is non-negotiable — merged canonicals are NOT counted as overlap candidates (their junction rows would have already been path-compressed to the terminal active id at merge time, but the filter is a defense-in-depth measure).
- Excludes the target episode itself.
- 80% test coverage threshold (per repo `vitest.config.ts`).
- `bun run lint && bun run test && bun run build` must pass.

### In Scope

- `computeCanonicalTopicOverlap` helper.
- `getCanonicalTopicOverlap` server action.
- Batch variant.
- Tests.

### Out of Scope

- UI integration (C2).
- Topic chips (C3).
- Topic detail page (D3).

## Dependencies

- Blocked by: #376 (EPIC A — canonical-topic foundation must ship first; specifically A2 schema + A5 pipeline so junction rows exist).
- Blocks: #395 (C2 UI integration).

## Definition of Done

- [ ] Helper + server action + batch variant + tests merged.
- [ ] Tests pass with ≥80% line coverage.
- [ ] PR approved and merged.
- [ ] Epic checklist updated.

### PR DESCRIPTION
**Pull request by rube-de**
## Summary

Adds canonical-topic-level overlap indicator (analog to ADR-034's category-level `getEpisodeTopicOverlap`, but operating on entity-resolved canonical topics for a far more precise signal). Implements the data layer only — UI integration ships in C2.

- **Pure helper** `computeCanonicalTopicOverlap` in `src/lib/topic-overlap.ts` — rows-in / result-out, no DB or async I/O. Mirrors the existing `computeTopicOverlap` split (pure logic in `/lib`, queries in actions).
- **Server actions** `getCanonicalTopicOverlap` (single) + `getCanonicalTopicOverlaps` (batch) in `src/app/actions/dashboard.ts`. Both wrapped in `withAuthAction`, both return `ActionResult<...>`. Single delegates to batch.
- **Up to 4 SQL queries per call** (Q3a/Q3b skipped when guards fire): target-canonicals JOIN `canonical_topics` WHERE `status='active'`; user-consumed episodes UNION (`listen_history` + `user_library`); per-canonical counts GROUP BY. Self-contribution subtracted (clamped at 0) when target is in user history.
- **Tie-break** consistent with existing canonical-topics convention: max count → highest `coverage_score` → lowest canonical id.
- **No `MIN_CONSUMED` gate** (intentional divergence from ADR-034 — canonical topics are entity-resolved and precise; a single overlap on "Opus 4.7" is a meaningful signal).
- **Shared `fetchConsumedEpisodeIds` helper** — extracted from the inlined UNION in `fetchUserTopicProfile` so ADR-034 and ADR-042 features can't drift on what counts as "consumed."
- **Capped client input** at `MAX_OVERLAP_LOOKUP_IDS = 500` with dedupe + `Array.isArray` guard, mirroring `listen-history.ts` precedent — bounds the SQL `IN` predicate against an untrusted client array.

### Decisions diverging from issue spec (each documented in plan)

| # | Issue says | We do | Why |
|---|-----------|-------|-----|
| 1 | DB-querying helper in `/lib/` | Pure helper + DB in action | Mirrors today's `computeTopicOverlap` / `fetchUserTopicProfile` split. |
| 2 | Action in `topics.ts` | Action in `dashboard.ts` | `topics.ts` is admin-only (every export uses `withAdminAction`). |
| 3 | `{ ok, data }` envelope | `ActionResult<T>` with `success` field | Codebase canonical (`src/types/action-result.ts`). |
| 4 | "status: unauthorized" | `"Unauthorized"` (capital U) | Matches `withAuthAction`. |
| 5 | (no mention of gates) | Skip `MIN_CONSUMED` gate | Canonical topics are precise per ADR-042; gate noise doesn't apply. |

Plan: `.dev/cdt/plans/plan-20260503-0003.md` (architect + PM debated; PM approved with one critical implementation note about Drizzle's `inArray(col, [])` throwing on empty arrays — three guards added).

Pre-PR multi-tool review (Codex + pr-review-toolkit + /simplify) flagged: §5/§6/§7 duplicate UNION query → extracted into shared helper; unbounded client input → capped + deduped; potential -1 from missing global count → `Math.max(0, …)` clamp; structured error logging; tightened JSDoc; missing tie-break test; missing self-subtraction-to-zero boundary test; missing DB-error test on the single wrapper. All addressed in commit `f827462`.

## Test plan

- [x] Unit tests for pure helper — 12 cases covering all branches (null, new, repeat, max-count-beats-higher-coverage, tie-breaks)
- [x] Action tests — 17 cases: anonymous → `Unauthorized`; empty input → `{}`; episode-not-found → `null`; no canonicals → `null`; all-zero → `'new'` with highest coverage; positive count → `'repeat'`; batch with mixed states; target-in-history exclusion; self-subtraction-to-zero boundary; missing-global-count clamp; all-unknown-ids batch (Q2 guard); zero-consumed user (Q3b guard); `status='active'` filter explicitly asserted; non-array input; DB error → graceful `{ success: false, error }`; single wrapper auth/found/not-found/error
- [x] Coverage: 100% lines on `src/lib/topic-overlap.ts` (helper); 84.46% global (above 80% threshold)
- [x] `bun run format:check`, `bun run lint`, `bun run build`, `bun run test` (2912 pass, 39 skipped) all pass; zero regressions

Closes #395.

### Open Questions
- **`getEpisodeTopicOverlap` `ActionResult` backfill** — sibling category-overlap action (`src/app/actions/dashboard.ts:461-509`) returns the raw result, not an `ActionResult`. Tracked in #441 (filed from this review).
- **C2 (UI integration) is out of scope** — tracked in #396; hand-off note posted there with action signatures and gotchas.
- **Design note (not a question, a hint for C2 / #396):** when `getCanonicalTopicOverlap` returns `null` (target has no canonicals linked), consumers should fall back to `getEpisodeTopicOverlap` (category). Recommendation-only — no consumer exists in this PR, so the fallback hasn't been validated end-to-end.
- **Q3a → subquery refactor** — efficiency reviewer flagged that pulling all consumed episode IDs into Node memory (`dashboard.ts:804`) before Q3b is unnecessary; could push consumed as a SQL subquery to drop one round trip and avoid serializing the list back into the Q3b `IN` predicate (`dashboard.ts:822`). Deferred per the reviewer's own note ("deserves its own PR with explain-plan validation").
- **`MAX_OVERLAP_LOOKUP_IDS = 500` cap is unvalidated** — picked to mirror `listen-history.ts` precedent. No batch consumer exists yet; revisit when the first caller lands so the cap can be sized to real usage instead of a guess. Inputs above the cap silently truncate (no error surfaced).

### Context for Next Session
- **`computeCanonicalTopicOverlap` is intentionally pure** — no DB imports, not async. DB queries live in the action layer. Mirrors the existing `computeTopicOverlap` / `fetchUserTopicProfile` split. If you're tempted to "simplify" by inlining DB calls into the helper, don't — testability suffers (helper is 100% covered with pure-function fixtures).
- **Three Drizzle empty-array guards are load-bearing** — `inArray(col, [])` throws at runtime in Drizzle. Guards exist after Q1, Q2, and Q3a. Without them, common cases (unknown episode IDs, brand-new user with no consumed history) raise an error instead of returning a graceful `null`. Do not remove these when refactoring.
- **Self-contribution subtraction with `Math.max(0, …)` clamp** — when target is in user history, Q3b counts the target itself; per-target count subtracts 1 then clamps to 0. The clamp guards against a race with reconcile-canonicals where a target's canonical row exists in Q2 but its global count is missing in Q3b — silently degrading to "new" is preferable to surfacing -1.
- **`fetchConsumedEpisodeIds` is the single source of truth** for "what counts as a consumed episode" (listen_history ∪ user_library). Both ADR-034 (category) and ADR-042 (canonical) overlap features call it. Adding a third source (e.g. dismissed/queued) updates both at once — do not inline a duplicate UNION.
- **No `merged_into_id` walking** — per ADR-042, canonicals are path-compressed at merge time. The `status='active'` filter on Q2 is the sole gate (defense-in-depth). Walking the merge chain is forbidden.
- **`MAX_OVERLAP_LOOKUP_IDS = 500` cap** — the action sanitizes client input (Array.isArray + filter + dedupe + slice). Without this cap, an unbounded array from the network would expand into the SQL `IN` predicate.

## Summary by CodeRabbit

* **New Features**
  * Added canonical-topic-level overlap indicators and endpoints to fetch overlap data (batch and single-item queries).

* **Tests**
  * Added extensive test coverage for canonical overlap selection, edge cases, input sanitization, auth flows, batching behaviors, and error handling.

* **Documentation**
  * Updated CHANGELOG with the new canonical-topic overlap entry.

### CODE REVIEW

File: src/app/actions/dashboard.ts [hunk 1]
```diff
@@ -668,3 +691,214 @@ export type RecentEpisode = PodcastIndexEpisode & {
...
+        .select({ id: episodes.id, podcastIndexId: episodes.podcastIndexId })
+        .from(episodes)
+        .where(inArray(episodes.podcastIndexId, sanitizedIds));
+
+      // Guard 1: no DB rows found → all inputs map to null
+      if (episodeRows.length === 0) {
+        const data = Object.fromEntries(
+          sanitizedIds.map((id) => [id, null]),
+        ) as Record<PodcastIndexEpisodeId, null>;
+        return { success: true, data };
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/app/actions/dashboard.ts]: Fixed: replaced `Object.fromEntries` with a null-prototype object (`Object.create(null)`) to prevent prototype-pollution via untrusted `__proto__` / `constructor` keys in the guard-1 null map.

File: src/app/actions/dashboard.ts [hunk 2]
```diff
@@ -668,3 +691,214 @@ export type RecentEpisode = PodcastIndexEpisode & {
...
+
+      const data = {} as Record<
+        PodcastIndexEpisodeId,
+        CanonicalOverlapResult | null
+      >;
+      for (const podcastIndexId of sanitizedIds) {
+        const dbId = podcastIndexToDbId.get(podcastIndexId);
+        if (dbId === undefined) {
+          data[podcastIndexId] = null;
+          continue;
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/app/actions/dashboard.ts]: Fixed: changed `data` accumulator from `{}` to `Object.create(null)` — direct property assignment with untrusted keys (e.g. `"__proto__"`) on a regular object triggers the prototype setter and pollutes `Object.prototype`; a null-prototype object is immune.

File: src/app/actions/dashboard.ts [hunk 3]
```diff
@@ -668,3 +691,214 @@ export type RecentEpisode = PodcastIndexEpisode & {
...
+              count: sql<number>`COUNT(DISTINCT ${episodeCanonicalTopics.episodeId})::integer`,
+            })
+            .from(episodeCanonicalTopics)
+            .where(
+              and(
+                inArray(
+                  episodeCanonicalTopics.canonicalTopicId,
+                  allCanonicalIds,
+                ),
+                inArray(episodeCanonicalTopics.episodeId, Array.from(consumed)),
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/app/actions/dashboard.ts]: Fixed: replaced `Array.from(consumed)` with `consumedIds` directly — `consumed` is populated solely from `consumedIds` at line 805 and no items are added between there and this call, so the conversion is redundant.

File: src/app/actions/dashboard.ts [hunk 4]
```diff
@@ -668,3 +691,213 @@ export type RecentEpisode = PodcastIndexEpisode & {
...
+        }
+      }
+
+      // Build result: per input episode, subtract self-contribution and call helper
+      const podcastIndexToDbId = new Map<PodcastIndexEpisodeId, number>();
+      for (const row of episodeRows) {
+        podcastIndexToDbId.set(row.podcastIndexId, row.id);
+      }
+
+      const data = Object.create(null) as Record<
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/app/actions/dashboard.ts]: Fixed: switched the `data` accumulator from `Object.create(null)` to plain `{}` so React Flight serialization at the server-action boundary no longer trips the "Only plain objects… null prototypes are not supported" runtime error when the action is invoked from a client component. The prototype-pollution surface that motivated the null-prototype object is now closed at the input boundary instead — `sanitizedIds` filters out `__proto__`/`constructor`/`prototype` upstream so they never reach object indexing. Both null-prototype sites (guard-1 path and the main result map) were updated; new tests cover the forbidden-key filter and assert the returned `data` has `Object.prototype`.

File: src/app/actions/dashboard.ts [hunk 5]
```diff
@@ -668,3 +691,213 @@ export type RecentEpisode = PodcastIndexEpisode & {
...
+    try {
+      const sanitizedIds = Array.isArray(podcastIndexEpisodeIds)
+        ? Array.from(
+            new Set(
+              podcastIndexEpisodeIds.filter(
+                (id): id is PodcastIndexEpisodeId =>
+                  typeof id === "string" && id.length > 0,
+              ),
+            ),
+          ).slice(0, MAX_OVERLAP_LOOKUP_IDS)
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/app/actions/dashboard.ts]: Fixed: ids are now trimmed before length validation so whitespace-only strings (`"   "`, `"\\n\\t"`) drop out, plus a `MAX_OVERLAP_ID_LENGTH = 256` per-id length bound and a forbidden-key filter (`__proto__`/`constructor`/`prototype`) round out the input sanitization. Mirrors the trim pattern in `recordListenEvent` (listen-history.ts:24) and the per-batch cap pattern in `getListenedEpisodeIds`. New tests cover trim, length-cap, and forbidden-key rejection.

File: src/app/actions/dashboard.ts [hunk 6]
```diff
@@ -668,3 +705,227 @@ export type RecentEpisode = PodcastIndexEpisode & {
...
+        const consumedIds = await fetchConsumedEpisodeIds(userId);
+        for (const id of consumedIds) consumed.add(id);
+
+        // Guard 3: no consumed episodes → skip Q3b (pass empty Map to helper)
+        if (consumed.size > 0) {
+          // Q3b: global per-canonical overlap counts
+          const countRows = await db
+            .select({
+              canonicalTopicId: episodeCanonicalTopics.canonicalTopicId,
+              count: sql<number>`COUNT(DISTINCT ${episodeCanonicalTopics.episodeId})::integer`,
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/app/actions/dashboard.ts]: Fixed: appended `.mapWith(Number)` to the Q3b count column to coerce the aggregate at selection time. Matches the codebase precedent in `topic-queries.ts:215`, `collections.ts:173`, and `resolution-metrics.ts:175`. Now `globalCounts` is always typed as `Map<number, number>` regardless of whether the driver returns the aggregate as string or number.

File: src/app/actions/dashboard.ts [hunk 7]
```diff
@@ -668,3 +705,227 @@ export type RecentEpisode = PodcastIndexEpisode & {
...
+  });
+}
+
+/** Single-episode convenience wrapper — delegates to the batch variant. */
+export async function getCanonicalTopicOverlap(
+  podcastIndexEpisodeId: PodcastIndexEpisodeId,
+): Promise<ActionResult<CanonicalOverlapResult | null>> {
+  const result = await getCanonicalTopicOverlaps([podcastIndexEpisodeId]);
+  if (!result.success) return result;
+  return { success: true, data: result.data[podcastIndexEpisodeId] ?? null };
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/app/actions/dashboard.ts]: Fixed: extracted `sanitizeOverlapId` (used by both the batch sanitization and this wrapper). The wrapper now normalizes the input first, delegates to the batch with the normalized id, and indexes the result map by that same key. Calls with recoverable variants like `" ep-42 "` no longer return `null` when the batch successfully computed an overlap under the trimmed key. Unusable inputs (whitespace-only, non-string, forbidden prototype keys) short-circuit to `{ success: true, data: null }` without an auth/DB hop.

[the remark this answers is not in the document]
Comment (rube-de) [reply src/app/actions/dashboard.ts]: Fixed: same fix as the Copilot thread immediately above. `getCanonicalTopicOverlap` now runs the input through the shared `sanitizeOverlapId` helper, passes the normalized value to the batch, and reads the result with the same key — so `" ep-42 "`/`"ep-42"` end up agreeing.

File: src/app/actions/dashboard.ts [hunk 8]
```diff
@@ -668,3 +722,234 @@ export type RecentEpisode = PodcastIndexEpisode & {
...
+ * (`"ep-42"`) while this wrapper read the raw key and returned `null`.
+ */
+export async function getCanonicalTopicOverlap(
+  podcastIndexEpisodeId: PodcastIndexEpisodeId,
+): Promise<ActionResult<CanonicalOverlapResult | null>> {
+  const normalized = sanitizeOverlapId(podcastIndexEpisodeId);
+  if (normalized === null) return { success: true, data: null };
+  const result = await getCanonicalTopicOverlaps([normalized]);
+  if (!result.success) return result;
+  return { success: true, data: result.data[normalized] ?? null };
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/app/actions/dashboard.ts]: Fixed: wrapped `getCanonicalTopicOverlap` in `withAuthAction`. Sanitization now happens inside the auth callback, so unauthenticated callers get `{ success: false, error: "Unauthorized" }` for any input — including unusable ones — instead of a success-shaped `null`. New test pins this contract for whitespace-only and forbidden-key inputs.

File: src/app/actions/dashboard.ts [hunk 9]
```diff
@@ -36,16 +38,64 @@ import {
...
+  if (typeof id !== "string") return null;
+  const trimmed = id.trim();
+  if (
+    trimmed.length === 0 ||
+    trimmed.length > MAX_OVERLAP_ID_LENGTH ||
+    FORBIDDEN_OVERLAP_KEYS.has(trimmed)
+  ) {
+    return null;
+  }
+  return trimmed as PodcastIndexEpisodeId;
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/app/actions/dashboard.ts]: Fixed: replaced `trimmed as PodcastIndexEpisodeId` with the codebase-standard `asPodcastIndexEpisodeId(trimmed)` constructor (already imported at the top of the file). Makes the post-validation cast intentional and consistent with `recordListenEvent` in listen-history.ts.

File: src/app/actions/dashboard.ts [hunk 10]
```diff
@@ -668,3 +724,238 @@ export type RecentEpisode = PodcastIndexEpisode & {
...
+  return withAuthAction(async (userId) => {
+    try {
+      const sanitizedIds: PodcastIndexEpisodeId[] = Array.isArray(
+        podcastIndexEpisodeIds,
+      )
+        ? Array.from(
+            new Set(
+              podcastIndexEpisodeIds
+                .map((id) => sanitizeOverlapId(id))
+                .filter((id): id is PodcastIndexEpisodeId => id !== null),
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/app/actions/dashboard.ts]: Fixed: replaced the `map().filter().Set` chain with a single-pass loop that early-exits at `MAX_OVERLAP_LOOKUP_IDS` (extracted as `sanitizeOverlapIdBatch`). A 10M-element untrusted input now stops being iterated after collecting 500 unique sanitized ids, instead of being fully mapped/filtered/deduped before `.slice()` kicked in. New test exercises a 600-element input with sentinel ids past the cap and asserts they never reach the result map.

File: src/app/actions/dashboard.ts [hunk 11]
```diff
@@ -668,3 +724,238 @@ export type RecentEpisode = PodcastIndexEpisode & {
...
+ * variant like `" ep-42 "` doesn't read a raw key the batch never computed.
+ */
+export async function getCanonicalTopicOverlap(
+  podcastIndexEpisodeId: PodcastIndexEpisodeId,
+): Promise<ActionResult<CanonicalOverlapResult | null>> {
+  return withAuthAction(async () => {
+    const normalized = sanitizeOverlapId(podcastIndexEpisodeId);
+    if (normalized === null) return { success: true, data: null };
+    const result = await getCanonicalTopicOverlaps([normalized]);
+    if (!result.success) return result;
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/app/actions/dashboard.ts]: Fixed: extracted `runCanonicalTopicOverlapBatch(userId, sanitizedIds)` as the impl helper. Both `getCanonicalTopicOverlaps` (batch) and `getCanonicalTopicOverlap` (single) wrap it with `withAuthAction` exactly once and call the impl directly — `auth()` runs once per request instead of twice for the single-episode path. New test asserts `mockAuth` is called exactly once.

File: src/app/actions/dashboard.ts [hunk 12]
```diff
@@ -55,9 +125,13 @@ async function fetchUserTopicProfile(userId: string) {
...
 
-  const totalConsumed = consumedRows.length;
-  const consumedIds = consumedRows.map((r) => r.episodeId);
+/** Fetch consumed episode IDs and build the user's topic profile in 2 batch queries. */
+async function fetchUserTopicProfile(userId: string) {
+  const consumedIds = await fetchConsumedEpisodeIds(userId);
+  const totalConsumed = consumedIds.length;
 
   let topicCountRows: Array<{ topic: string; count: number }> = [];
   if (consumedIds.length > 0) {
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/app/actions/dashboard.ts:148]: Fixed: appended `.mapWith(Number)` to the `fetchUserTopicProfile` count column too — same Neon-HTTP-string-aggregate guard already applied to Q3b in the canonical-overlap path. `buildUserTopicProfile` now always receives numeric counts.

File: src/app/actions/__tests__/dashboard.test.ts [hunk 13]
```diff
@@ -1351,3 +1379,658 @@ describe("getTrendingTopicBySlug", () => {
...
+      success: true,
+      data: null,
+    });
+    expect(
+      await getCanonicalTopicOverlap(null as unknown as PodcastIndexEpisodeId),
+    ).toEqual({
+      success: true,
+      data: null,
+    });
+    // None of these should hit the auth/DB layers.
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/app/actions/__tests__/dashboard.test.ts]: Fixed: corrected the comment to say auth runs (the wrapper is auth-first) but DB queries are skipped because sanitization short-circuits inside the `withAuthAction` callback before any DB call. Added `expect(mockAuth).toHaveBeenCalled()` so the assertion matches the corrected intent.

File: src/app/actions/dashboard.ts [hunk 14]
```diff
@@ -36,16 +38,84 @@ import {
...
+// Stopping mid-iteration bounds CPU/memory regardless of input length —
+// applying `.slice(MAX, ...)` only AFTER `.map().filter().Set` would still
+// process the full untrusted client array, which is a DoS surface.
+function sanitizeOverlapIdBatch(input: unknown): PodcastIndexEpisodeId[] {
+  if (!Array.isArray(input)) return [];
+  const seen = new Set<string>();
+  const out: PodcastIndexEpisodeId[] = [];
+  for (const raw of input) {
+    if (out.length >= MAX_OVERLAP_LOOKUP_IDS) break;
+    const id = sanitizeOverlapId(raw);
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/app/actions/dashboard.ts]: Fixed: added `MAX_OVERLAP_INSPECT_IDS = 2000` (4× the unique-id cap) as a hard cap on raw inputs inspected. The sanitizer now bails out either when it has collected 500 unique valid ids OR after inspecting 2000 raw entries — whichever fires first. Closes the gap where a payload of millions of duplicates / invalid strings would walk the full array because the unique-id cap never fires. New test: 10000 whitespace-only inputs return `{}` and never reach the DB.

File: src/app/actions/dashboard.ts [hunk 15]
```diff
@@ -36,16 +38,84 @@ import {
...
+    if (id === null || seen.has(id)) continue;
+    seen.add(id);
+    out.push(id);
+  }
+  return out;
```
[the remark this answers is not in the document]
Comment (rube-de) [reply src/app/actions/dashboard.ts]: Fixed: added `MAX_OVERLAP_INSPECT_IDS = 2000` (4× the unique-id cap) as a hard cap on raw inputs inspected. The sanitizer now bails out either when it has collected 500 unique valid ids OR after inspecting 2000 raw entries — whichever fires first. Closes the gap where a payload of millions of duplicates / invalid strings would walk the full array because the unique-id cap never fires. New test: 10000 whitespace-only inputs return `{}` and never reach the DB.
